### PR TITLE
Export untagged content items

### PR DESCRIPTION
**Pull request by koetsier**
Saves the complete in-scope dump from the content store
Adds functions to export data from all untagged items to a file.

Trello: 
- https://trello.com/c/2C792Ubu/53-identify-in-scope-content-not-tagged-to-any-taxon
- https://trello.com/c/sIgcxfgU/130-changes-in-data-extraction-pipeline

### CODE REVIEW

File: python/data_extraction/export_data.py [hunk 1]
```diff
@@ -25,6 +25,17 @@
                                 content_store_url=plek.find('draft-content-store'))
 
 
+untagged_content_links_generator = content_export.content_links_generator(
+    additional_search_fields={'reject_part_of_taxonomy_tree': taxonomy_query.TaxonomyQuery().root_taxon_content_ids()},
```
Comment (tomsabin) [inline python/data_extraction/export_data.py]: I think there is some ambiguity in the card requirements, and you may have clarified this already but I'll bring it up anyway in the hope that you can clarify it for me. The Trello description has as follows

> We should identify all in-scope content items not associated with a taxon

My understanding of this is that a taxon can be anywhere, and not just attached to the 'topic taxonomy' (i.e. the one built up here from the level_one_taxons). With the current implementation, we might have a problem that the World taxons would be included, which might not be what we're after.

File: python/data_extraction/export_data.py [hunk 2]
```diff
@@ -40,6 +51,16 @@ def content():
     return filter(lambda link: link, map(slicer, content_generator))
 
 
+def untagged_content():
+    pool = Pool(10)
```
Comment (tomsabin) [inline python/data_extraction/export_data.py]: My knowledge of multithreaded programming is limited, but how did you decide on 10 processes for this? I'm assuming it should depend on the underlying hardware that will be running this.

File: python/data_extraction/export_data.py (hunk 1 above)
Comment (koetsier) [reply python/data_extraction/export_data.py]: Good point. I thought this would be correct - I'll ask for some clarification

Comment (koetsier) [reply python/data_extraction/export_data.py]: Fixed this now.

File: python/data_extraction/export_data.py [hunk 3]
```diff
@@ -9,17 +9,33 @@
...
+
 get_content = functools.partial(content_export.get_content,
-                                base_fields=configuration['base_fields'],
-                                taxon_fields=configuration['taxon_fields'],
-                                ppo_fields=configuration['ppo_fields'],
                                 content_store_url=plek.find('draft-content-store'))
 
 
+untagged_content_links_generator = content_export.content_links_generator(
+    additional_search_fields={'filter_part_of_taxonomy_tree': '_MISSING'},
```
Comment (cbaines) [inline python/data_extraction/export_data.py]: As discussed, this isn't ideal as Rummager doesn't have the associations of non Whitehall content with draft taxons.

Comment (koetsier) [reply python/data_extraction/export_data.py]: fixed that in a pretty comprehensive rewrite

File: Makefile [hunk 4]
```diff
@@ -64,7 +71,7 @@ clean :
 	    $(DATADIR)/labelled.csv.gz  $(DATADIR)/filtered.csv.gz  $(DATADIR)/old_taxons.csv.gz  \
 	    $(DATADIR)/labelled_level1.csv.gz  $(DATADIR)/labelled_level2.csv.gz  \
 	    $(DATADIR)/empty_taxons_not_world.csv.gz  $(DATADIR)/new_content.csv.gz \
-	    data/taxons.json data/content.json
+	    data/taxons.json data/content.json.gz data/export_untagged_content.json.gz data/export_untagged_content.json.gz
```
Comment (cbaines) [inline Makefile]: `data/export_untagged_content.json.gz` looks to be here twice, does it need to be?

File: python/data_extraction/content_export.py [hunk 5]
```diff
@@ -1,11 +1,17 @@
...
 
-def content_links_generator(page_size=1000, rummager_url=plek.find('rummager'), blacklist_document_types=[]):
-    search_results = rummager.Rummager(rummager_url).search_generator(
-        {'reject_content_store_document_type': blacklist_document_types, 'fields': ['link']}, page_size=page_size)
+
+def content_links_generator(page_size=1000,
+                            additional_search_fields = {},
+                            rummager_url=plek.find('rummager'),
+                            blacklist_document_types=[]):
+    search_hash = merge({'reject_content_store_document_type': blacklist_document_types,
```
Comment (cbaines) [inline python/data_extraction/content_export.py]: This isn't technically hash, we should try to keep the Rubyisms to a minimum.

File: python/data_extraction/export_data.py [hunk 6]
```diff
@@ -1,63 +1,78 @@
...
+with open(config_path) as json_data_file:
     configuration = json.load(json_data_file)
 
 
-content_links_generator = content_export.content_links_generator(
-    blacklist_document_types=configuration['blacklist_document_types']
-)
+def __transform_content(input_filename="data/content.json.gz",
+                        output_filename="data/filtered_content.json.gz",
+                        transform_function=lambda x: x):
```
Comment (cbaines) [inline python/data_extraction/export_data.py]: I'm not quite sure about the custom JSON parsing stuff, but I like what you've done here with transforming the file to generate a new file.

File: python/data_extraction/export_data.py [hunk 7]
```diff
@@ -1,63 +1,78 @@
...
+        with gzip.open(output_filename, mode='wt') as output_file:
+            content_generator = json_arrays.read_json(input_file)
+            json_arrays.write_json(output_file, transform_function(content_generator))
 
 
-slicer = functools.partial(content_export.content_hash_slicer,
-                           base_fields=configuration['base_fields'],
-                           taxon_fields=configuration['taxon_fields'],
-                           ppo_fields=configuration['ppo_fields'])
+def export_content(output_filename="data/content.json"):
```
Comment (cbaines) [inline python/data_extraction/export_data.py]: It looks like this commit is changing the format to be gzipped always. If so, the default filename should probably be `"data/content.json.gz"`.

File: python/data_extraction/export_data.py [hunk 8]
```diff
@@ -1,63 +1,78 @@
...
 
-def taxonomy():
-    query = taxonomy_query.TaxonomyQuery()
-    level_one_taxons = query.level_one_taxons()
-    children = [taxon
-                for level_one_taxon in level_one_taxons
-                for taxon in query.child_taxons(level_one_taxon['base_path'])]
-    return iter(level_one_taxons + children)
 
+def export_filtered_content(input_filename="data/content.json.gz", output_filename="data/filtered_content.json.gz"):
```
Comment (cbaines) [inline python/data_extraction/export_data.py]: It would be useful if the name for this file made it clear what it contains. What is the filtering in this case?

File: python/data_extraction/content_export.py [hunk 9]
```diff
@@ -22,6 +28,20 @@ def content_hash_slicer(content_hash, base_fields=[], taxon_fields=[], ppo_field
     return result
 
 
+def untagged_hash_slicer(content_hash, base_fields=[], ppo_fields=[]):
```
Comment (cbaines) [inline python/data_extraction/content_export.py]: Maybe say `untagged_content_slicer`, or `untagged_dict_slicer` as `hash` is not a thing in Python.

File: python/data_extraction/taxonomy_query.py [hunk 10]
```diff
@@ -9,6 +9,9 @@ def __init__(self, key_list=("content_id", "base_path", "title"),
         self.content_store_url = content_store_url
         self.key_list = key_list
 
+    def root_taxon_content_ids(self):
+        return [root_taxon.get('content_id') for root_taxon in self.level_one_taxons()]
```
Comment (cbaines) [inline python/data_extraction/taxonomy_query.py]: The "level one" naming is being used in part here. I think it would be clearer to use that everywhere, so call this `level_one_taxon_content_ids`. What do you think?

Comment (cbaines) [reply python/data_extraction/taxonomy_query.py]: Ah, I'm not even sure this function is used. At least it doesn't appear to be used in this commit.

File: python/test/data_extraction/test_export_data.py [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+            self.assertEqual(expected, json.loads(output.buffer))
+
+    def export_untagged_content(self):
+        output = MockIO()
+        input_string = "[" + json.dumps(content_with_taxons) + ",\n" + json.dumps(content_without_taxons) + "\n]"
+
+        with unittest.mock.patch('gzip.open',
+                                 side_effect=self.return_input_or_output(io.StringIO(input_string), output)):
+            export_data.export_filtered_content()
+            self.assertListEqual([content_without_taxons], json.loads(output.buffer))
```
Comment (cbaines) [inline python/test/data_extraction/test_export_data.py]: These tests look fine to me. But, I'm uncertain about the value of testing in this way. My hope is that we'll be able to exercise run the full data extraction process regularly, which should give confidence that it works, and also run it when making changes to ensure things don't break.

Comment (koetsier) [reply python/test/data_extraction/test_export_data.py]: The data extraction process takes a long time and isn't being run as part of the unit tests. My feeling is that the code that these tests exercises is complex enough to warrant some unit tests.

File: python/data_extraction/content_export.py (hunk 5 above)
Comment (koetsier) [reply python/data_extraction/content_export.py]: renamed all 'hash' to 'dict'

File: python/data_extraction/content_export.py (hunk 9 above)
Comment (koetsier) [reply python/data_extraction/content_export.py]: renamed hash->dict

File: python/data_extraction/taxonomy_query.py (hunk 10 above)
Comment (koetsier) [reply python/data_extraction/taxonomy_query.py]: removed this

File: python/data_extraction/export_data.py (hunk 6 above)
Comment (koetsier) [reply python/data_extraction/export_data.py]: json is hard to parse incrementally - i.e. without reading in the whole file in one go and then keeping the entire content-store in memory

File: python/data_extraction/export_data.py (hunk 7 above)
Comment (koetsier) [reply python/data_extraction/export_data.py]: fixed

File: python/data_extraction/export_data.py (hunk 8 above)
Comment (koetsier) [reply python/data_extraction/export_data.py]: tricky to say because it may change. It filters a bunch of attributes of the dictionary - what those are is what you define in the dict_slicer

File: Makefile [hunk 12]
```diff
@@ -13,12 +13,19 @@ taxons : $(DATADIR)/clean_taxons.csv.gz
...
+data/content.json.gz:
+	cd python && python3 -u -c "from data_extraction.export_data import export_content; export_content(output_filename='../data/content.json.gz')"
+
+data/export_filtered_content.json.gz : data/content.json.gz
+	cd python && python3 -u -c "from data_extraction.export_data import export_filtered_content; export_filtered_content(input_filename='../data/content.json.gz', output_filename='../data/filtered_content.json.gz')"
+
+data/export_untagged_content.json.gz : data/content.json.gz
+	cd python && python3 -u -c "from data_extraction.export_data import export_untagged_content; export_untagged_content(input_filename='../data/content.json.gz', output_filename='../data/untagged_content.json.gz')"
 
 data/taxons.json:
```
Comment (cbaines) [inline Makefile]: This should be data/taxons.json.gz, now that the file it generates has changed.

File: python/data_extraction/export_data.py (hunk 6 above)
Comment (cbaines) [reply python/data_extraction/export_data.py]: There look to be some libraries that talk about parsing JSON in a stream, maybe we can use one of them. I've used streaming XML parsers in the past, so hopefully the same thing exists for JSON. It would be good to have a solution that is independent of the formatting of the file. This is ok as a starting point though.

File: Makefile (hunk 12 above)
Comment (cbaines) [reply Makefile]: I've amended the relevant commit now.

**Comment by cbaines** [comment]
Looks like this worked when it ran in integration https://deploy.integration.publishing.service.gov.uk/job/govuk_taxonomy_supervised_learning/17/console
